test/jdk/java/io/InputStreamReader/StatefulDecoderNearEOF.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8292043
+ * @run testng StatefulDecoderNearEOF
+ * @summary Check MalformedInputException is thrown with stateful decoders
+ *      with malformed input before EOF
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.CodingErrorAction;
+import java.nio.charset.MalformedInputException;
+import java.nio.charset.StandardCharsets;
+
+import org.testng.annotations.Test;
+import static org.testng.Assert.assertThrows;
+
+@Test
+public class StatefulDecoderNearEOF {
+    private static final byte[] INPUT = {
+            (byte) 0xff, (byte) 0xfe, // BOM (in UTF-16LE)
+            0, (byte) 0xd8, // High surrogate (in UTF-16LE)
+    };
+
+    public void testStatefulDecoderNearEOF() {
+        assertThrows(MalformedInputException.class, () -> {
+            try (var r = new InputStreamReader(
+                    new ByteArrayInputStream(INPUT),
+                    StandardCharsets.UTF_16.newDecoder().onMalformedInput(CodingErrorAction.REPORT))) {
+                System.out.printf("%04x%n", r.read()); // \u00d8 (wrong, uses UTF-16BE)
+                System.out.printf("%04x%n", r.read()); // EOF
```

Comment:
This will pass if either read fails, I think the test should be checking that the first call to read throws MalformedInputException.

Would it be feasible to add a second test where there are characters between the BOM and the truncated high surrogate? It would be possible to decode those characters before it fails at the end of the stream.

Reply:
Addressed both points.